File: app/src/main/java/com/thewizrd/simpleweather/widgets/WeatherWidgetProvider4x1Notification.java
package com.thewizrd.simpleweather.widgets;

import android.appwidget.AppWidgetManager;
import android.content.ComponentName;
import android.content.Context;

import com.thewizrd.simpleweather.App;
import com.thewizrd.simpleweather.R;

import static com.thewizrd.simpleweather.widgets.WidgetType.Widget4x1Notification;

public class WeatherWidgetProvider4x1Notification extends WeatherWidgetProvider {
    private static WeatherWidgetProvider4x1Notification sInstance;

    public static synchronized WeatherWidgetProvider4x1Notification getInstance() {
        if (sInstance == null)
            sInstance = new WeatherWidgetProvider4x1Notification();

        return sInstance;
    }

    // Overrides
    @Override
    public WidgetType getWidgetType() {
        return Widget4x1Notification;
    }

    @Override
    public int getWidgetLayoutId() {
        return R.layout.app_widget_4x1_notification;
    }

    @Override
    protected String getClassName() {
        return this.getClass().getName();
    }

    @Override
    public ComponentName getComponentName() {
        return new ComponentName(App.getInstance().getAppContext(), getClassName());
    }

    @Override
    public boolean hasInstances(Context context) {
        AppWidgetManager appWidgetManager = AppWidgetManager.getInstance(context);
        int[] appWidgetIds = appWidgetManager.getAppWidgetIds(
                new ComponentName(context, getClassName()));
        return (appWidgetIds.length > 0);
    }
}

File: app/src/main/java/com/thewizrd/simpleweather/widgets/WeatherWidgetProvider4x1Notification.kt
package com.thewizrd.simpleweather.widgets

import com.thewizrd.simpleweather.R

class WeatherWidgetProvider4x1Notification : WeatherWidgetProvider() {
    override val info: WidgetProviderInfo by lazy { Widget4x1NotificationInfo() }
}

class Widget4x1NotificationInfo : WidgetProviderInfo() {
    // Overrides
    override val widgetType: WidgetType
        get() = WidgetType.Widget4x1Notification
    override val widgetLayoutId: Int
        get() = R.layout.app_widget_4x1_notification

    companion object {
        private var _instance: WidgetProviderInfo? = null

        @JvmStatic
        fun getInstance(): WidgetProviderInfo {
            if (_instance == null) {
                _instance = Widget4x1NotificationInfo()
            }
            return _instance!!
        }
    }
}